File: src/user_examples/cs_user_source_terms-richards.f90
!-------------------------------------------------------------------------------

!VERS

! This file is part of Code_Saturne, a general-purpose CFD tool.
!
! Copyright (C) 1998-2018 EDF S.A.
!
! This program is free software; you can redistribute it and/or modify it under
! the terms of the GNU General Public License as published by the Free Software
! Foundation; either version 2 of the License, or (at your option) any later
! version.
!
! This program is distributed in the hope that it will be useful, but WITHOUT
! ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
! FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
! details.
!
! You should have received a copy of the GNU General Public License along with
! this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
! Street, Fifth Floor, Boston, MA 02110-1301, USA.

!-------------------------------------------------------------------------------

!===============================================================================
! Purpose:
! -------

!> \file cs_user_source_terms-richards.f90
!>
!> \brief User source terms example.
!>
!> See \subpage cs_user_source_terms and \subpage cs_user_source_terms-scalar_in_a_channel
!> for examples.
!>
!> \brief Additional right-hand side source terms for velocity components equation
!> (Navier-Stokes)
!>
!> \section ustsnv_use  Usage
!>
!> The additional source term is decomposed into an explicit part (\c crvexp) and
!> an implicit part (\c crvimp) that must be provided here.
!> The resulting equation solved by the code for a velocity is:
!> \f[
!>  \rho \norm{\vol{\celli}} \DP{\vect{u}} + ....
!>   = \tens{crvimp} \cdot \vect{u} + \vect{crvexp}
!> \f]
!>
!> Note that \c crvexp and \c crvimp are defined after the Finite Volume integration
!> over the cells, so they include the "volume" term. More precisely:
!>   - crvexp is expressed in kg.m/s2
!>   - crvimp is expressed in kg/s
!>
!> The \c crvexp and \c crvimp arrays are already initialized to 0
!> before entering the
!> the routine. It is not needed to do it in the routine (waste of CPU time).
!>
!> \remark The additional force on \f$ x_i \f$ direction is given by
!>  \c crvexp(i, iel) + vel(j, iel)* crvimp(j, i).
!>
!> For stability reasons, Code_Saturne will not add -crvimp directly to the
!> diagonal of the matrix, but Max(-crvimp,0). This way, the crvimp term is
!> treated implicitely only if it strengthens the diagonal of the matrix.
!> However, when using the second-order in time scheme, this limitation cannot
!> be done anymore and -crvimp is added directly. The user should therefore test
!> the negativity of crvimp by himself.
!>
!> When using the second-order in time scheme, one should supply:
!>   - crvexp at time n
!>   - crvimp at time n+1/2
!>
!> The selection of cells where to apply the source terms is based on a
!> \ref getcel command. For more info on the syntax of the \ref getcel command,
!> refer to the user manual or to the comments on the similar command
!> \ref getfbr in the routine \ref cs_user_boundary_conditions.
!
!-------------------------------------------------------------------------------

subroutine ustssc &
!================

 ( nvar   , nscal  , ncepdp , ncesmp ,                            &
   iscal  ,                                                       &
   icepdc , icetsm , itypsm ,                                     &
   dt     ,                                                       &
   ckupdc , smacel ,                                              &
   crvexp , crvimp )

!===============================================================================
! Module files
!===============================================================================

use paramx
use numvar
use entsor
use optcal
use cstphy
use parall
use period
use mesh
use field
use darcy_module
use cs_c_bindings

!===============================================================================

implicit none

integer          nvar   , nscal
integer          ncepdp , ncesmp
integer          iscal

integer          icepdc(ncepdp)
integer          icetsm(ncesmp), itypsm(ncesmp,nvar)

double precision dt(ncelet)
double precision ckupdc(6,ncepdp), smacel(ncesmp,nvar)
double precision crvexp(ncelet), crvimp(ncelet)

character(len=80) ::  chaine

integer          iiscvr,  iel
integer          icelt, ncelt
integer, allocatable, dimension(:) :: lstcel

double precision lambda, leaching_time, leaching_volume
double precision, dimension(:), pointer :: delay, saturation

type(var_cal_opt) :: vcopt

!===============================================================================

allocate(lstcel(ncel))

call field_get_label(ivarfl(isca(iscal)), chaine)

iiscvr = iscavr(iscal)

call field_get_key_struct_var_cal_opt(ivarfl(isca(iscal)), vcopt)

if (vcopt%iwarni.ge.1) then
  write(nfecra,1000) chaine(1:8)
endif

!======================
! Leaching
!======================

!< [richards_leaching]
! Leaching from volumic zone labelled as 'LEACHING_ZONE'
! Set the duration of leaching
leaching_time = 1.d2

call getcel('LEACHING_ZONE', ncelt, lstcel)

! Compute the volume of the leaching zone
leaching_volume = 0.d0
do icelt = 1, ncelt
  iel = lstcel(icelt)
  leaching_volume = leaching_volume + volume(iel)
enddo

if (irangp.ge.0) then
  call parsom(leaching_volume)
endif

! Compute the source term
do icelt = 1, ncelt
  iel = lstcel(icelt)
  !Progressive leaching
  if (ttcabs.lt.leaching_time) then
    crvexp(iel) = 1. / leaching_time * volume(iel) / leaching_volume &
         * exp(-lambda*ttcabs) / saturation(iel) / delay(iel)
  else
    crvexp(iel) = 0.d0
  endif
enddo
!< [richards_leaching]


 1000 format(' User source terms for variable ',A8,/)

deallocate(lstcel)

return
end subroutine ustssc

File: src/user_examples/cs_user_source_terms-richards.c
/*============================================================================
 * User source terms example for the Richards equation.
 *============================================================================*/

/* VERS */

/*
  This file is part of code_saturne, a general-purpose CFD tool.

  Copyright (C) 1998-2022 EDF S.A.

  This program is free software; you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
  Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

/*----------------------------------------------------------------------------*/

#include "cs_defs.h"

/*----------------------------------------------------------------------------
 * Standard C library headers
 *----------------------------------------------------------------------------*/

#include <assert.h>
#include <math.h>

/*----------------------------------------------------------------------------
 * PLE library headers
 *----------------------------------------------------------------------------*/

#include <ple_coupling.h>

/*----------------------------------------------------------------------------
 * Local headers
 *----------------------------------------------------------------------------*/

#include "cs_headers.h"

/*----------------------------------------------------------------------------*/

BEGIN_C_DECLS

/*----------------------------------------------------------------------------*/
/*!
 * \file cs_user_source_terms-richards.c
 *
 * \brief User source terms example for the Richards equation.
 *
 * See the reference \ref cs_user_source_terms.c for documentation.
 */
/*----------------------------------------------------------------------------*/

/*============================================================================
 * User function definitions
 *============================================================================*/

/*----------------------------------------------------------------------------*/
/*!
 * \brief Function called at each time step to define source terms.
 *
 * \param[in, out]  domain   pointer to a cs_domain_t structure
 * \param[in]       f_id     field id of the variable
 * \param[out]      st_exp   explicit source term
 * \param[out]      st_imp   implicit part of the source term
 */
/*----------------------------------------------------------------------------*/

void
cs_user_source_terms(cs_domain_t  *domain,
                     int           f_id,
                     cs_real_t    *st_exp,
                     cs_real_t    *st_imp)
{
  CS_NO_WARN_IF_UNUSED(domain);
  CS_NO_WARN_IF_UNUSED(st_imp);

  /*! [richards_leaching_init] */
  /* local variables */

  const cs_field_t  *f = cs_field_by_id(f_id);
  const cs_real_t  t = cs_glob_time_step->t_cur;

  const cs_real_t  *cell_f_vol = cs_glob_mesh_quantities->cell_vol;

  /* Map saturation and delay */

  char f_name[64];
  snprintf(f_name, 63, "%s_delay", f->name); f_name[63] = '\0';
  const cs_real_t  *delay = cs_field_by_name(f_name)->val;

  const cs_real_t  *saturation = cs_field_by_name("saturation")->val;
  /*! [richards_leaching_init] */

  /* logging */
  /*! [richards_leaching_log] */
  const cs_equation_param_t *eqp = cs_field_get_equation_param_const(f);

  if (eqp->verbosity >= 1)
    bft_printf(" User source terms for variable %s\n",
               cs_field_get_label(f));
  /*! [richards_leaching_log] */

  /* Leaching from volume zone labeled as "LEACHING_ZONE" */

  /*! [richards_leaching] */
  cs_real_t lambda = 1.e-2;        /* first order decay coefficient */
  cs_real_t leaching_time = 1.e2;  /* leaching duration */

  const cs_zone_t *z = cs_volume_zone_by_name_try("LEACHING_ZONE");

  /* progressive leaching */

  if (z != NULL && t < leaching_time) {
    cs_real_t leaching_volume = z->f_measure;

    for (cs_lnum_t i = 0; i < z->n_elts; i++) {
      cs_lnum_t c_id = z->elt_ids[i];

      st_exp[c_id] =   cell_f_vol[i] * exp(-lambda*t)
                     / (  leaching_time * leaching_volume
                        * saturation[i] * delay[i]);
    }
  }
  /*! [richards_leaching] */
}

/*----------------------------------------------------------------------------*/

END_C_DECLS
